js/pro/alpaca.js

```diff
@@ -0,0 +1,715 @@
+    }
+
+    async watchOrderBook (symbol, limit = undefined, params = {}) {
+        /**
+         * @method
+         * @name alpaca#watchOrderBook
+         * @description watches information on open orders with bid (buy) and ask (sell) prices, volumes and other data
+         * @param {string} symbol unified symbol of the market to fetch the order book for
+         * @param {int|undefined} limit the maximum amount of order book entries to return.
+         * @param {object} params extra parameters specific to the alpaca api endpoint
+         * @returns {object} A dictionary of [order book structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-book-structure} indexed by market symbols
+         */
+        const url = this.urls['api']['ws']['crypto'];
+        await this.authenticate (url);
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        symbol = market['symbol'];
+        const messageHash = 'orderbook' + ':' + symbol;
+        const request = {
+            'action': 'subscribe',
+            'orderbooks': [ market['id'] ],
+        };
+        const orderbook = await this.watch (url, messageHash, this.extend (request, params), messageHash);
+        return orderbook.limit ();
+    }
+
+    handleOrderBook (client, message) {
+        //
+        // snapshot
+        //    [{
+        //        T: "o",
+        //        S: "BTC/USDT",
+        //        t: "2022-12-16T06:35:31.585113205Z",
+        //        b: [{
+        //                p: 17394.37,
+        //                s: 0.015499,
+        //            },
+        //            ...
+        //        ],
+        //        a: [{
+        //                p: 17398.8,
+        //                s: 0.042919,
+        //            },
+        //            ...
+        //        ],
+        //        r: true,
+        //    }]
+        //
+        const first = this.safeValue (message, 0, {});
+        const marketId = this.safeString (first, 'S');
+        const symbol = this.safeSymbol (marketId);
+        const datetime = this.safeString (first, 't');
+        const timestamp = this.iso8601 (datetime);
+        const isSnapshot = this.safeValue (first, 'r', false);
+        let orderbook = this.safeValue (this.orderbooks, symbol);
+        if (orderbook === undefined) {
+            orderbook = this.orderBook ();
+        }
+        if (isSnapshot) {
+            const snapshot = this.parseOrderBook (first, symbol, timestamp, 'b', 'a', 'p', 's');
+            orderbook.reset (snapshot);
+        } else {
+            const asks = this.safeValue (first, 'a', []);
+            const bids = this.safeValue (first, 'b', []);
+            this.handleDeltas (orderbook['asks'], asks);
+            this.handleDeltas (orderbook['bids'], bids);
+            orderbook['timestamp'] = timestamp;
+            orderbook['datetime'] = this.iso8601 (timestamp);
+        }
+        const messageHash = 'orderbook' + ':' + symbol;
```

Comment:
@pcriadoperez something's wrong here and crashes in python because `symbol` will be None
![image](https://user-images.githubusercontent.com/43336371/208423071-b7f1c734-f03f-43e5-ae76-23bc5a75bd77.png)